logstash-core/src/test/java/org/logstash/config/ir/CompiledPipelineTest.java

```diff
@@ -560,4 +563,150 @@ public Filter buildFilter(final String name, final String id,
             return null;
         }
     }
+
+    @Test
+    @SuppressWarnings({"unchecked"})
+    public void testCompilerCacheCompiledClasses() throws IOException, InvalidIRException {
+        ComputeStepSyntaxElement.cleanClassCache();
+        final FixedPluginFactory pluginFactory = new FixedPluginFactory(
+                () -> null,
+                () -> IDENTITY_FILTER,
+                mockOutputSupplier()
+        );
+
+        final PipelineIR pipeline1 = ConfigCompiler.configToPipelineIR(
+                IRHelpers.toSourceWithMetadataFromPath("org/logstash/config/ir/cache/pipeline1.conf"),
+                false);
+        final CompiledPipeline compiledPipeline1 = new CompiledPipeline(pipeline1, pluginFactory);
+        compiledPipeline1.buildExecution();
+        final int cachedBefore = ComputeStepSyntaxElement.classCacheSize();
+
+        final PipelineIR pipeline2 = ConfigCompiler.configToPipelineIR(
+                IRHelpers.toSourceWithMetadataFromPath("org/logstash/config/ir/cache/pipeline2.conf"),
+                false);
+        final CompiledPipeline compiledPipeline2 = new CompiledPipeline(pipeline2, pluginFactory);
+        compiledPipeline2.buildExecution();
+        final int cachedAfter = ComputeStepSyntaxElement.classCacheSize();
+        final String message = String.format("unexpected cache size, cachedAfter: %d, cachedBefore: %d", cachedAfter, cachedBefore);
+        assertEquals(message, 1, cachedAfter - cachedBefore);
```

Comment:
Think back to rspec tests and how the framework gives us `let()` and other tools to remove preparation content from the test itself, I'd suggest grouping together the setup and then the test. Example:

```suggestion
        final FixedPluginFactory pluginFactory = new FixedPluginFactory(
                () -> null,
                () -> IDENTITY_FILTER,
                mockOutputSupplier()
        );

        final PipelineIR baselinePipeline = ConfigCompiler.configToPipelineIR(
                IRHelpers.toSourceWithMetadataFromPath("org/logstash/config/ir/cache/pipeline1.conf"),
                false);
        final CompiledPipeline cBaselinePipeline = new CompiledPipeline(baselinePipeline, pluginFactory);

        final PipelineIR pipelineWithExtraFilter = ConfigCompiler.configToPipelineIR(
                IRHelpers.toSourceWithMetadataFromPath("org/logstash/config/ir/cache/pipeline2.conf"),
                false);
        final CompiledPipeline cPipelineWithExtraFilter = new CompiledPipeline(pipelineWithExtraFilter, pluginFactory);
        
        // actual test: compiling a pipeline with an extra filter should only create 1 extra class
        ComputeStepSyntaxElement.cleanClassCache();
        cBaselinePipeline.buildExecution();
        final int cachedBefore = ComputeStepSyntaxElement.classCacheSize();
        cPipelineWithExtraFilter.buildExecution();
        final int cachedAfter = ComputeStepSyntaxElement.classCacheSize();
        
        final String message = String.format("unexpected cache size, cachedAfter: %d, cachedBefore: %d", cachedAfter, cachedBefore);
        assertEquals(message, 1, cachedAfter - cachedBefore);
```